ratex/csrc/compiler/raf_node_lowering.cpp

```diff
@@ -994,6 +998,117 @@ Var RAFNodeLowering::LowerMean(const ir::ops::Mean* node) {
   return BuildMean({x}, node);
 }
 
+Var BuildMatMul(const std::vector<Var>& ops, const ir::ops::MatMul* node) {
+  LTC_CHECK_EQ(node->operands().size(), 2) << "Unexpected number of operands";
+  Var x = ops[0];
+  Var y = ops[1];
+  Var mm;
+  std::vector<int64_t> a_shape = node->a_shape();
+  std::vector<int64_t> b_shape = node->b_shape();
+  int64_t a_size = a_shape.size();
+  int64_t b_size = b_shape.size();
+
+  if (a_size > 2 && b_size > 2) {
+    if (a_size != 3) {
+      x = BindSymbol(raf::ir::Call(
+          Op::Get("raf.op.reshape"),
+          {x, MakeConstant(TupleInt({-1, a_shape[a_size - 2], a_shape[a_size - 1]}))}));
+    }
+
+    auto trans_axes = lazy_tensors::util::Iota<int64_t>(b_size);
+    std::swap(trans_axes[b_size - 2], trans_axes[b_size - 1]);
+    y = BindSymbol(
+        raf::ir::Call(Op::Get("raf.op.transpose"), {y, MakeConstant(TupleInt(trans_axes))}));
```

Comment:
We should have all 4 variants (raf.op.batch_matmul_nn, raf.op.batch_matmul_nt, etc) so you should be able to get rid of transpose in all cases?

Reply:
I think we will might need to create separate ops for each case. Originally I had a type variable in MatMul node which we can pass nn/nt/tt and concat in node_lowering. But the conversions (reshaping/transposes/etc) do not hold for all.

Reply:
I don't get the point. What's the issue of creating separate ops for each case?

Reply:
Sounds good, I can create separate ops for each case, but I will make separate PRs for those.

Reply:
Sorry if I miss anything. I'd prefer to correct it in this PR, as the current implementation introduces unnecessary overhead but doesn't improve the readability IMHO.

Reply:
IIUC, the problem is that in PyTorch, matmul accepts all these shapes combination and does the reshape and transpose underneath. Awais is trying to mimic what it does in PyTorch and XLA.

Reply:
That makes sense. We did the same thing in Relay in the past so this is totally correct.
I'm just wondering if there's any urgent task that needs this PR. If so then it's fine to merge first and fix later; otherwise, I'd prefer to polish it before merging.

Reply:
I understand your point, and in fact, originally we had removed removed the trans_axes and used batch_matmul. But while running test cases for bert like/transformer style models we got errors which were resolved by this. Despite this, as we discussed yesterday, even though Pytorch does not have matmul_nt/tn/tt, we should be able to take advantage of RAF capability via Ratex (this is especially useful in auto diff). So I added a custom matmul_xx OP for these cases. The op is only prototype for the meantime (sufficient for bert/transformers) only acting as a bridge, but it will require its own algorithm (reshapes/transforms) to handle all possible cases. I will make an issue for this.